File: 3_data_fits/surrogate.jl
module surrogate_models
export Surrogate, train, basis, predict, Guassian, scaling

using LinearAlgebra
import LinearAlgebra.norm

abstract type AbstractSurrogate{T} end

mutable struct LS{T} <: AbstractSurrogate{T}
    X::Matrix{T}
    Y::Matrix{T}
    lb::Vector{T}
    ub::Vector{T}
    scale::Bool
    r::T
    d::Int
    n_centers::Int
    dim_i::Int
    dim_o::Int
    isempty::Bool
    B::Matrix{T}
    W::Matrix{T}
    J::Matrix{T}

    function LS(::Type{T}=Float64; X::Matrix{T}, Y::Matrix{T}, lb::Vector{T}, ub::Vector{T}, r::T, d::Int, scale::Bool) where T
        
        ls = new{T}()
        ls.X = X
        ls.Y = Y
        ls.r = r
        ls.d = d
        ls.lb = lb
        ls.ub = ub
        ls.scale = scale

        # Dimensionality
        ls.n_centers = size(X,1)
        ls.dim_i = size(X,2)
        ls.dim_o = size(Y,2)

        # Weights
        ls.isempty = true
        ls.B = zeros(T,(size(X,1),size(X,2)*d+1))
        ls.W = zeros(T,(size(X,2)*d+1,size(Y,2)))

        # reguralization
        ls.J = I(size(X,2)*d+1) .* r
        ls.J[1,1] = 0
        return ls
    end
end

function Polynomial(d::Int, ζ::Vector{T}) where T
    n = length(ζ)
    b = zeros(T, d*n+1)
    b[1] = 1
    for i = 1:d
        b[2+(i-1)*n:1+i*n] = ζ.^i
    end
    return b
end

function basis(d::Int, Z::Matrix{T}, f::Function) where T

    B = zeros(T, size(Z, 1), size(Z,2)*d+1)

    for k = 1:size(Z,1)
        B[k,:] = f(d,Z[k,:])
    end
    return B
end

function train(m::LS)
    B = basis(m.d,m.X,Polynomial)
    m.B = B
    m.W = pinv(B' * B + m.J) * B' * m.Y
    m.isempty = false
end

function predict(m::LS{T}, Z::Matrix{T}) where T

    if size(Z,2) != size(m.X,2)
        throw(ArgumentError("dimenions of prediction and training sites are not equal"))
    end

    if m.isempty
        throw(DomainError(m,"model is not trained!"))
    end

    if m.scale
        Z = scaling(Z,m.lb,m.ub,1)
    end
    
    basis(m.d,Z,Polynomial)*m.W
end

mutable struct RBF{T} <: AbstractSurrogate{T}
    X::Matrix{T}
    Y::Matrix{T}
    lb::Vector{T}
    ub::Vector{T}
    scale::Bool
    r::T
    λ::T
    kernel::Function
    n_centers::Int
    dim_i::Int
    dim_o::Int
    isempty::Bool
    B::Matrix{T}
    W::Matrix{T}
    J::Matrix{T}

    function RBF(::Type{T}=Float64; X::Matrix{T}, Y::Matrix{T}, lb::Vector{T}, ub::Vector{T}, r::T, λ::T, kernel::Function, scale::Bool) where T
        
        rbf = new{T}()
        rbf.X = X
        rbf.Y = Y
        rbf.lb = lb
        rbf.ub = ub
        rbf.scale = scale
        rbf.r = r
        rbf.λ = λ
        rbf.kernel = kernel
        
        # Dimensionality
        rbf.n_centers = size(X,1)
        rbf.dim_i = size(X,2)
        rbf.dim_o = size(Y,2)

        # Weights
        rbf.isempty = true
        rbf.B = zeros(T,(size(X,1),size(X,1)))
        rbf.W = zeros(T,(size(X,1),size(Y,2)))

        # reguralization
        rbf.J = I(size(X,1)) .* r
        rbf.J[1,1] = 0
        return rbf
    end
end

# basis functions
function norm(X::Matrix{T}, dims::Integer) where T
    sqrt.(sum(X.^2, dims=dims))
end

function Guassian(λ::T, ζ::Vector{T}, X::Matrix{T}) where T
    exp.(-λ*norm(ζ' .- X, 2).^2)
end

function basis(λ::T, Z::Matrix{T}, X::Matrix{T}, f::Function) where T

    B = zeros(T, size(Z, 1), size(X,1))

    for k = 1:size(Z,1)
        B[k,:] = f(λ,Z[k,:],X)
    end
    return B
end

function train(m::RBF)
    B = basis(m.λ,m.X,m.X,m.kernel)
    m.B = B
    m.W = pinv(B' * B + m.J) * B' * m.Y
    m.isempty = false
end

function predict(m::RBF{T}, Z::Matrix{T}) where T

    if size(Z,2) != size(m.X,2)
        throw(ArgumentError("dimenions of prediction and training sites are not equal"))
    end

    if m.isempty
        throw(DomainError(m,"model is not trained!"))
    end

    if m.scale
        Z = scaling(Z,m.lb,m.ub,1)
    end

    basis(m.λ,Z,m.X,m.kernel)*m.W
end
    
mutable struct Surrogate{T}
    X::Matrix{T}
    Y::Matrix{T}
    name::String
    type::String
    model::AbstractSurrogate{T}

    function Surrogate(::Type{T}=Float64; X::Matrix{T}, Y::Matrix{T}, type::String, 
        lb::Union{Vector{T}, Nothing}=nothing, ub::Union{Vector{T}, Nothing}=nothing, 
        r::T=1e-6, λ::T=15.0, d::Int=1, kernel::Function=Guassian, scale::Bool=true, 
        name::String = "") where T

        if size(X,1) != size(Y,1)
            throw(ArgumentError("number of training inputs does not match number of training outputs"))
        end

        if isnothing(lb)
            lb = vec(minimum(X,dims=1))
        end

        if isnothing(ub)
            ub = vec(maximum(X,dims=1))
        end

        if scale
            X = scaling(X,lb,ub,1)
        end

        sur = new{T}()
        sur.type = type
        sur.name = name

        if type == "RBF"
            sur.model = RBF(T, X=X , Y=Y, lb=lb, ub=ub, r=r, λ=λ, kernel=kernel, scale=scale)
        end

        if type == "LS"
            sur.model = LS(T, X=X , Y=Y, lb=lb, ub=ub, r=r, d=d, scale=scale)
        end

        return sur
    end
end

function scaling(X::Matrix{T}, lb::Vector{T}, ub::Vector{T}, type::Integer)::Matrix{T} where T
    if type == 1
        Xs = (X' .- lb)' ./ (ub - lb)'
    elseif type == 2
        Xs =  X .* (ub - lb)' .+ lb'
    end
    return Xs
end

end # module

File: 3_data_fits/surrogate.py
import numpy as np
from abc import ABC, abstractmethod
from scipy.linalg import pinv
from numpy.linalg import norm

def scaling(X, lb, ub, type):
    if type == 1:
        return (X - lb) / (ub - lb)
    elif type == 2:
        return X * (ub - lb) + lb

class AbstractSurrogate(ABC):
    def __init__(self):
        pass

    @abstractmethod
    def train(self):
        pass

    @abstractmethod
    def predict(self, Z):
        pass

class LS(AbstractSurrogate):
    def __init__(self, X, Y, lb, ub, r, d, scale):
        self.X = X
        self.Y = Y
        self.lb = lb
        self.ub = ub
        self.r = r
        self.d = d
        self.scale = scale
        self.n_centers, self.dim_i = X.shape
        self.dim_o = Y.shape[1]
        self.isempty = True
        self.B = np.zeros((self.n_centers, self.dim_i*self.d+1))
        self.W = np.zeros((self.dim_i*self.d+1, self.dim_o))
        self.J = np.eye(self.dim_i*self.d+1) * r
        self.J[0,0] = 0

    def Polynomial(self, d, ζ):
        n = len(ζ)
        b = np.ones((d*n+1))
        for i in range(0, d):
            b[1+i*n:(i+1)*n+1] = ζ**(i+1)
        return b

    def basis(self, d, Z):
        return np.array([self.Polynomial(d, z) for z in Z])

    def train(self):
        self.B = self.basis(self.d, self.X)
        self.W = pinv(self.B.T @ self.B + self.J) @ self.B.T @ self.Y
        self.isempty = False

    def predict(self, Z):
        if Z.shape[1] != self.X.shape[1]:
            raise ValueError("Dimensions of prediction and training sites are not equal")

        if self.isempty:
            raise ValueError("Model is not trained!")

        if self.scale:
            Z = scaling(Z, self.lb, self.ub, 1)

        return self.basis(self.d, Z) @ self.W

class RBF(AbstractSurrogate):
    def __init__(self, X, Y, lb, ub, r, λ, kernel, scale):
        self.X = X
        self.Y = Y
        self.lb = lb
        self.ub = ub
        self.scale = scale
        self.r = r
        self.λ = λ
        self.kernel = kernel
        self.n_centers, self.dim_i = X.shape
        self.dim_o = Y.shape[1]
        self.isempty = True
        self.B = np.zeros((self.n_centers, self.n_centers))
        self.W = np.zeros((self.n_centers, self.dim_o))
        self.J = np.eye(self.n_centers) * r
        self.J[0,0] = 0

    def Gaussian(self, λ, ζ, X):
        return np.exp(-λ * norm(ζ - X,axis=1))[:,None]

    def basis(self, λ, Z):
        return np.hstack([self.Gaussian(λ, z, self.X) for z in Z])

    def train(self):
        self.B = self.basis(self.λ, self.X)
        self.W = pinv(self.B.T @ self.B + self.J) @ self.B.T @ self.Y
        self.isempty = False

    def predict(self, Z):
        if Z.shape[1] != self.X.shape[1]:
            raise ValueError("Dimensions of prediction and training sites are not equal")

        if self.isempty:
            raise ValueError("Model is not trained!")

        if self.scale:
            Z = scaling(Z, self.lb, self.ub, 1)

        return self.basis(self.λ, Z).T @ self.W

class Surrogate:
    def __init__(self, X, Y, type, lb=None, ub=None, r=1e-6, λ=15.0, d=1, kernel=None, scale=True, name=""):
        if X.shape[0] != Y.shape[0]:
            raise ValueError("Number of training inputs does not match number of training outputs")

        if lb is None:
            lb = np.min(X, axis=0)

        if ub is None:
            ub = np.max(X, axis=0)

        if scale:
            X = scaling(X, lb, ub, 1)

        self.name = name
        self.type = type

        if type == "RBF":
            self.model = RBF(X, Y, lb, ub, r, λ, kernel, scale)
        elif type == "LS":
            self.model = LS(X, Y, lb, ub, r, d, scale)

    def train(self):
        self.model.train()

    def predict(self, Z):
        return self.model.predict(Z)
